helix-term/src/ui/menu.rs

```diff
@@ -15,6 +15,11 @@ use fuzzy_matcher::FuzzyMatcher;
 use helix_view::{graphics::Rect, Editor};
 use tui::layout::Constraint;
 
+pub enum SortStrategy {
+    Text,
+    Score,
+}
+
```

Comment:
I'm a bit wary of adding more strategies. How about `Item` implements `fn score(matcher: &...)` and the default implementation just calls `matcher.fuzzy_match`? Then to override for preselected items, the `Item` trait implementation for CompletionItem could return `0` if preselected.

Reply:
I'd also remove the `preselected` feature and just always pre-select.

Reply:
I removed the `preselect` config option to always use the LSP preselected item. Is this what you meant by your second comment ?

However, I am not quite sure I follow with the first one ! I don't see how having `Item` implement `fn score(matcher: &...)` would help get rid of the `SortStrategy` enum.

I use the `SortStrategy` enum at the menu level. It was necessary because as of now, the `Menu` only sorts alphabetically (the code to sort by score is commented out in master). This makes sense for code actions (the only other instance of `Menu` in use), but not for completion items that need to be sorted according to their fuzzy match score. As of now, completions items are just filtered (matches the pattern or not) and are then sorted alphabetically.
Moreover, the score given to an item depends of the prompt pattern being empty or not, so it cannot be contained in the `Item` trait implementation. That way, when the completion menu pops open, the user sees the preselected items first, but then when they type anything, they see the items sorted by score.

Reply:
Ah I see, so this is a change unrelated to preselect/preview doc (and is there to fix code action sorting)? Can you split it out into a separate PR to discuss there instead?

> Moreover, the score given to an item depends of the prompt pattern being empty or not, so it cannot be contained in the Item trait implementation.

You could pass in both the matcher and the current pattern into the scoring function.

Reply:
Alright, I just opened #3215 for the sorting strategy. I'll rebase this one on master once merged 😉

Reply:
Oh I mean I hoped to merge this PR without requiring to land the sort strategy. If preselect depends on it then yeah it's OK to review together

Reply:
At least it'll make a clearer diff once rebased, because with rustfmt reindenting big portions of the code, the few modifications to add sort-strategies made a huge diff.